[CoolMasterNet] devices don't update - errors in log when binding restarts

There was a previous [issue](https://github.com/openhab/openhab2-addons/issues/3644) (solved) with the binding stopping updating devices, and errors when the binding restarted.

There now seems to be another similar error. Same symptoms - binding will communicate outbound commands to the coolmasternet but stops updating state changes to openhab. Different log errors - see below.

Many thanks!

```
2018-07-08 21:32:18.939 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler HVACHandler of thing coolmasternet:hvac:main:kidsbed tried updating channel set_temp although the handler was already disposed.
2018-07-08 21:32:19.188 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler HVACHandler of thing coolmasternet:hvac:main:living tried updating channel on although the handler was already disposed.
2018-07-08 21:32:19.188 [ERROR] [ome.core.thing.link.ThingLinkManager] - Exception occurred while informing handler: null
java.lang.NullPointerException: null
	at org.openhab.binding.coolmasternet.handler.HVACHandler.getControllerHandler(HVACHandler.java:52) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.query(HVACHandler.java:122) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.refresh(HVACHandler.java:103) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.handleCommand(HVACHandler.java:81) ~[?:?]
	at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.channelLinked(BaseThingHandler.java:278) ~[?:?]
	at org.eclipse.smarthome.core.thing.link.ThingLinkManager.lambda$0(ThingLinkManager.java:300) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
2018-07-08 21:32:19.442 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler HVACHandler of thing coolmasternet:hvac:main:masterbed tried updating channel current_temp although the handler was already disposed.
2018-07-08 21:32:19.443 [ERROR] [ome.core.thing.link.ThingLinkManager] - Exception occurred while informing handler: null
java.lang.NullPointerException: null
	at org.openhab.binding.coolmasternet.handler.HVACHandler.getControllerHandler(HVACHandler.java:52) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.query(HVACHandler.java:122) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.refresh(HVACHandler.java:104) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.handleCommand(HVACHandler.java:81) ~[?:?]
	at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.channelLinked(BaseThingHandler.java:278) ~[?:?]
	at org.eclipse.smarthome.core.thing.link.ThingLinkManager.lambda$0(ThingLinkManager.java:300) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
2018-07-08 21:32:19.693 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler HVACHandler of thing coolmasternet:hvac:main:masterbed tried updating channel fan_speed although the handler was already disposed.
2018-07-08 21:32:19.943 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler HVACHandler of thing coolmasternet:hvac:main:masterbed tried updating channel fan_speed although the handler was already disposed.
2018-07-08 21:32:20.201 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler HVACHandler of thing coolmasternet:hvac:main:living tried updating channel current_temp although the handler was already disposed.
2018-07-08 21:32:20.203 [ERROR] [ome.core.thing.link.ThingLinkManager] - Exception occurred while informing handler: null
java.lang.NullPointerException: null
	at org.openhab.binding.coolmasternet.handler.HVACHandler.getControllerHandler(HVACHandler.java:52) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.query(HVACHandler.java:122) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.refresh(HVACHandler.java:104) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.handleCommand(HVACHandler.java:81) ~[?:?]
	at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.channelLinked(BaseThingHandler.java:278) ~[?:?]
	at org.eclipse.smarthome.core.thing.link.ThingLinkManager.lambda$0(ThingLinkManager.java:300) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
2018-07-08 21:33:02.256 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler HVACHandler of thing coolmasternet:hvac:main:living tried updating channel on although the handler was already disposed.
2018-07-08 21:33:15.524 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler HVACHandler of thing coolmasternet:hvac:main:kidsbed tried updating channel on although the handler was already disposed.
2018-07-08 21:33:15.783 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler HVACHandler of thing coolmasternet:hvac:main:masterbed tried updating channel on although the handler was already disposed.
2018-07-08 21:33:15.784 [ERROR] [ome.core.thing.link.ThingLinkManager] - Exception occurred while informing handler: null
java.lang.NullPointerException: null
	at org.openhab.binding.coolmasternet.handler.HVACHandler.getControllerHandler(HVACHandler.java:52) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.query(HVACHandler.java:122) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.refresh(HVACHandler.java:99) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.handleCommand(HVACHandler.java:81) ~[?:?]
	at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.channelLinked(BaseThingHandler.java:278) ~[?:?]
	at org.eclipse.smarthome.core.thing.link.ThingLinkManager.lambda$0(ThingLinkManager.java:300) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
2018-07-08 21:33:15.784 [ERROR] [ome.core.thing.link.ThingLinkManager] - Exception occurred while informing handler: null
java.lang.NullPointerException: null
	at org.openhab.binding.coolmasternet.handler.HVACHandler.getControllerHandler(HVACHandler.java:52) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.query(HVACHandler.java:122) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.refresh(HVACHandler.java:103) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.handleCommand(HVACHandler.java:81) ~[?:?]
	at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.channelLinked(BaseThingHandler.java:278) ~[?:?]
	at org.eclipse.smarthome.core.thing.link.ThingLinkManager.lambda$0(ThingLinkManager.java:300) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
2018-07-08 21:33:16.024 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler HVACHandler of thing coolmasternet:hvac:main:masterbed tried updating channel louvre_angle although the handler was already disposed.
2018-07-08 21:33:16.025 [ERROR] [ome.core.thing.link.ThingLinkManager] - Exception occurred while informing handler: null
java.lang.NullPointerException: null
	at org.openhab.binding.coolmasternet.handler.HVACHandler.getControllerHandler(HVACHandler.java:52) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.query(HVACHandler.java:122) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.refresh(HVACHandler.java:113) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.handleCommand(HVACHandler.java:81) ~[?:?]
	at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.channelLinked(BaseThingHandler.java:278) ~[?:?]
	at org.eclipse.smarthome.core.thing.link.ThingLinkManager.lambda$0(ThingLinkManager.java:300) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
2018-07-08 21:33:16.275 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler HVACHandler of thing coolmasternet:hvac:main:kidsbed tried updating channel on although the handler was already disposed.
2018-07-08 21:33:16.275 [ERROR] [ome.core.thing.link.ThingLinkManager] - Exception occurred while informing handler: null
java.lang.NullPointerException: null
	at org.openhab.binding.coolmasternet.handler.HVACHandler.getControllerHandler(HVACHandler.java:52) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.query(HVACHandler.java:122) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.refresh(HVACHandler.java:103) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.handleCommand(HVACHandler.java:81) ~[?:?]
	at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.channelLinked(BaseThingHandler.java:278) ~[?:?]
	at org.eclipse.smarthome.core.thing.link.ThingLinkManager.lambda$0(ThingLinkManager.java:300) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
2018-07-08 21:33:16.523 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler HVACHandler of thing coolmasternet:hvac:main:masterbed tried updating channel current_temp although the handler was already disposed.
2018-07-08 21:33:16.523 [ERROR] [ome.core.thing.link.ThingLinkManager] - Exception occurred while informing handler: null
java.lang.NullPointerException: null
	at org.openhab.binding.coolmasternet.handler.HVACHandler.getControllerHandler(HVACHandler.java:52) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.query(HVACHandler.java:122) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.refresh(HVACHandler.java:104) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.handleCommand(HVACHandler.java:81) ~[?:?]
	at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.channelLinked(BaseThingHandler.java:278) ~[?:?]
	at org.eclipse.smarthome.core.thing.link.ThingLinkManager.lambda$0(ThingLinkManager.java:300) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
```

Oh dear, possibly another problem with the binding. Just got this error, after which I think things have stopped updating:

```
2018-07-12 18:56:08.182 [ERROR] [ng.coolmasternet.handler.HVACHandler] - Query 'query L1.101 o' failed: No response from CoolMasterNet unit 192.168.1.130:10102
2018-07-12 18:56:09.393 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception
java.lang.NumberFormatException: null
	at java.math.BigDecimal.<init>(BigDecimal.java:494) ~[?:?]
	at java.math.BigDecimal.<init>(BigDecimal.java:383) ~[?:?]
	at java.math.BigDecimal.<init>(BigDecimal.java:806) ~[?:?]
	at org.eclipse.smarthome.core.library.types.DecimalType.<init>(DecimalType.java:54) ~[?:?]
	at org.openhab.binding.coolmasternet.handler.HVACHandler.refresh(HVACHandler.java:110) ~[?:?]
	at org.openhab.binding.coolmasternet.internal.ControllerHandler.refreshHVACUnits(ControllerHandler.java:100) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
```

--- a/addons/binding/org.openhab.binding.coolmasternet/src/main/java/org/openhab/binding/coolmasternet/handler/HVACHandler.java
+++ b/addons/binding/org.openhab.binding.coolmasternet/src/main/java/org/openhab/binding/coolmasternet/handler/HVACHandler.java
@@ -18,6 +18,7 @@ import org.eclipse.jdt.annotation.Nullable;
 import org.eclipse.smarthome.core.library.types.DecimalType;
 import org.eclipse.smarthome.core.library.types.OnOffType;
 import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.thing.Bridge;
 import org.eclipse.smarthome.core.thing.ChannelUID;
 import org.eclipse.smarthome.core.thing.Thing;
 import org.eclipse.smarthome.core.thing.ThingStatus;
@@ -38,6 +39,7 @@ import org.slf4j.LoggerFactory;
  * HVAC unit (a single UID on a CoolMasterNet controller.)
  *
  * @author Angus Gratton - Initial contribution
+ * @author Wouter Born - Fix null pointer exceptions
  */
 @NonNullByDefault
 public class HVACHandler extends BaseThingHandler {
@@ -49,7 +51,8 @@ public class HVACHandler extends BaseThingHandler {
     }
 
     private @Nullable ControllerHandler getControllerHandler() {
-        return (ControllerHandler) getBridge().getHandler();
+        Bridge bridge = getBridge();
+        return bridge != null ? (ControllerHandler) bridge.getHandler() : null;
     }
 
     @Override
@@ -59,8 +62,12 @@ public class HVACHandler extends BaseThingHandler {
         ControllerHandler controller = getControllerHandler();
 
         try {
-            if (controller == null || !controller.isConnected()) {
-                ControllerConfiguration config = getBridge().getConfiguration().as(ControllerConfiguration.class);
+            Bridge bridge = getBridge();
+            if (bridge == null) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                        "CoolMasterNet Controller bridge not set");
+            } else if (controller == null || !controller.isConnected()) {
+                ControllerConfiguration config = bridge.getConfiguration().as(ControllerConfiguration.class);
                 updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
                         String.format("Could not connect to CoolMasterNet unit %s:%d", config.host, config.port));
             } else {
@@ -88,7 +95,7 @@ public class HVACHandler extends BaseThingHandler {
 
     @Override
     public void initialize() {
-        logger.debug("Initialising CoolMasterNet HVAC handler...");
+        logger.debug("Initializing CoolMasterNet HVAC handler...");
         updateStatus(ThingStatus.ONLINE);
     }
 
@@ -124,7 +131,7 @@ public class HVACHandler extends BaseThingHandler {
             try {
                 return controller.sendCommand(command);
             } catch (CoolMasterClientError e) {
-                logger.error("Query {} failed: {}", command, e.getMessage());
+                logger.error("Query '{}' failed: {}", command, e.getMessage());
             }
         }
         return null; /* passing back null sets an invalid value on the channel */

--- a/addons/binding/org.openhab.binding.coolmasternet/src/main/java/org/openhab/binding/coolmasternet/internal/ControllerHandler.java
+++ b/addons/binding/org.openhab.binding.coolmasternet/src/main/java/org/openhab/binding/coolmasternet/internal/ControllerHandler.java
@@ -43,16 +43,18 @@ import org.slf4j.LoggerFactory;
  * These are individual Things inside the bridge.
  *
  * @author Angus Gratton - Initial contribution
+ * @author Wouter Born - Fix null pointer exceptions and stop refresh job on update/dispose
  */
 @NonNullByDefault
 public class ControllerHandler extends BaseBridgeHandler {
     private static final int SOCKET_TIMEOUT = 2000;
+
     private final Logger logger = LoggerFactory.getLogger(ControllerHandler.class);
-    private @Nullable String host;
-    private int port;
-    private @Nullable Socket socket;
-    private final Object lock = new Object();
+    private final Object refreshLock = new Object();
+    private final Object socketLock = new Object();
+
     private @Nullable ScheduledFuture<?> refreshJob;
+    private @Nullable Socket socket;
 
     public ControllerHandler(Bridge thing) {
         super(thing);
@@ -60,35 +62,49 @@ public class ControllerHandler extends BaseBridgeHandler {
 
     @Override
     public void initialize() {
-        logger.debug("Initialising CoolMasterNet Controller handler...");
-
-        ControllerConfiguration config = getConfigAs(ControllerConfiguration.class);
-        host = config.host;
-        port = config.port;
-
-        Runnable refreshHVACUnits = () -> {
-            try {
-                checkConnection();
-                updateStatus(ThingStatus.ONLINE);
-                for (Thing t : getThing().getThings()) {
-                    HVACHandler h = (HVACHandler) t.getHandler();
-                    if (h != null) {
-                        h.refresh();
-                    }
-                }
-            } catch (CoolMasterClientError e) {
-                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
-            }
-        };
-        refreshJob = scheduler.scheduleWithFixedDelay(refreshHVACUnits, 0, config.refresh, TimeUnit.SECONDS);
+        logger.debug("Initializing CoolMasterNet Controller handler...");
+        stopRefresh();
+        startRefresh();
     }
 
     @Override
     public void dispose() {
-        if (refreshJob != null) {
-            refreshJob.cancel(true);
-        }
+        stopRefresh();
         super.dispose();
+    }
+
+    private void startRefresh() {
+        synchronized (refreshLock) {
+            ControllerConfiguration config = getConfigAs(ControllerConfiguration.class);
+            logger.debug("Scheduling new refresh job");
+            refreshJob = scheduler.scheduleWithFixedDelay(this::refreshHVACUnits, 0, config.refresh, TimeUnit.SECONDS);
+        }
+    }
+
+    private void stopRefresh() {
+        synchronized (refreshLock) {
+            ScheduledFuture<?> localRefreshJob = refreshJob;
+            if (localRefreshJob != null && !localRefreshJob.isCancelled()) {
+                logger.debug("Cancelling existing refresh job");
+                localRefreshJob.cancel(true);
+                refreshJob = null;
+            }
+        }
+    }
+
+    private void refreshHVACUnits() {
+        try {
+            checkConnection();
+            updateStatus(ThingStatus.ONLINE);
+            for (Thing t : getThing().getThings()) {
+                HVACHandler h = (HVACHandler) t.getHandler();
+                if (h != null) {
+                    h.refresh();
+                }
+            }
+        } catch (CoolMasterClientError e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+        }
     }
 
     /*
@@ -98,8 +114,9 @@ public class ControllerHandler extends BaseBridgeHandler {
      * and try to re-establish the connection if possible.
      */
     public boolean isConnected() {
-        synchronized (this.lock) {
-            return socket != null && socket.isConnected() && !socket.isClosed();
+        synchronized (socketLock) {
+            Socket localSocket = socket;
+            return localSocket != null && localSocket.isConnected() && !localSocket.isClosed();
         }
     }
 
@@ -109,18 +126,24 @@ public class ControllerHandler extends BaseBridgeHandler {
      * If the "OK" prompt is not received then a CoolMasterClientError is thrown that contains whatever
      * error message was printed by the CoolMasterNet.
      */
+    @SuppressWarnings("resource")
     public @Nullable String sendCommand(String command) throws CoolMasterClientError {
-        synchronized (this.lock) {
+        synchronized (socketLock) {
             checkConnection();
 
             StringBuilder response = new StringBuilder();
             try {
+                Socket localSocket = socket;
+                if (localSocket == null || !isConnected()) {
+                    throw new CoolMasterClientError(String.format("No connection for sending command %s", command));
+                }
+
                 logger.trace("Sending command '{}'", command);
-                OutputStream out = socket.getOutputStream();
+                OutputStream out = localSocket.getOutputStream();
                 out.write(command.getBytes());
                 out.write("\r\n".getBytes());
 
-                BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+                BufferedReader in = new BufferedReader(new InputStreamReader(localSocket.getInputStream()));
                 while (true) {
                     String line = in.readLine();
                     logger.trace("Read result '{}'", line);
@@ -153,25 +176,33 @@ public class ControllerHandler extends BaseBridgeHandler {
      * May block for 1-2 seconds.
      *
      * Throws CoolMasterNetClientError if there is a connection problem.
-     *
      */
-    public void checkConnection() throws CoolMasterClientError {
-        synchronized (this.lock) {
+    @SuppressWarnings("resource")
+    private void checkConnection() throws CoolMasterClientError {
+        synchronized (socketLock) {
+            ControllerConfiguration config = getConfigAs(ControllerConfiguration.class);
             try {
                 if (!isConnected()) {
                     connect();
                     if (!isConnected()) {
-                        throw new CoolMasterClientError(String.format("Failed to connect to %s:%s", host, port));
+                        throw new CoolMasterClientError(
+                                String.format("Failed to connect to %s:%s", config.host, config.port));
                     }
                 }
 
-                InputStream in = socket.getInputStream();
+                Socket localSocket = socket;
+                if (localSocket == null) {
+                    throw new CoolMasterClientError(
+                            String.format("Failed to connect to %s:%s", config.host, config.port));
+                }
+
+                InputStream in = localSocket.getInputStream();
                 /* Flush anything pending in the input stream */
                 while (in.available() > 0) {
                     in.read();
                 }
                 /* Send a CRLF, expect a > prompt (and a CRLF) back */
-                OutputStream out = socket.getOutputStream();
+                OutputStream out = localSocket.getOutputStream();
                 out.write("\r\n".getBytes(StandardCharsets.US_ASCII));
                 /*
                  * this will time out with IOException if it doesn't see that prompt
@@ -183,36 +214,42 @@ public class ControllerHandler extends BaseBridgeHandler {
                 }
             } catch (IOException e) {
                 disconnect();
-                logger.error("{}", e.getLocalizedMessage(), e);
-                throw new CoolMasterClientError(String.format("No response from CoolMasterNet unit %s:%s", host, port));
+                logger.debug("{}", e.getLocalizedMessage(), e);
+                throw new CoolMasterClientError(
+                        String.format("No response from CoolMasterNet unit %s:%s", config.host, config.port));
             }
         }
     }
 
     private void connect() throws IOException {
-        synchronized (this.lock) {
+        synchronized (socketLock) {
+            ControllerConfiguration config = getConfigAs(ControllerConfiguration.class);
             try {
-                socket = new Socket();
-                socket.connect(new InetSocketAddress(host, port), SOCKET_TIMEOUT);
-                socket.setSoTimeout(SOCKET_TIMEOUT);
+                Socket localSocket = new Socket();
+                localSocket.connect(new InetSocketAddress(config.host, config.port), SOCKET_TIMEOUT);
+                localSocket.setSoTimeout(SOCKET_TIMEOUT);
+                socket = localSocket;
             } catch (UnknownHostException e) {
-                logger.error("unknown socket host {}", host);
+                logger.error("Unknown socket host: {}", config.host);
                 socket = null;
             } catch (SocketException e) {
-                logger.error("{}", e.getLocalizedMessage(), e);
+                logger.error("Failed to connect to {}:{}: {}", config.host, config.port, e.getLocalizedMessage(), e);
                 socket = null;
             }
         }
     }
 
-    public void disconnect() {
-        synchronized (this.lock) {
-            try {
-                socket.close();
-            } catch (IOException e1) {
-                logger.error("{}", e1.getLocalizedMessage(), e1);
+    private void disconnect() {
+        synchronized (socketLock) {
+            Socket localSocket = socket;
+            if (localSocket != null) {
+                try {
+                    localSocket.close();
+                } catch (IOException e1) {
+                    logger.error("{}", e1.getLocalizedMessage(), e1);
+                }
+                socket = null;
             }
-            socket = null;
         }
     }
 

